### PR TITLE
Lower ldmatrix as kir::Asm

### PR DESCRIPTION
**Pull request by zasdfgbnm**
In `kir::Asm`, we no longer need to manually specify the argument list in its ctor. Instead, the argument list will be generated on the fly, for example:
```diff
  IrBuilder::create<kir::Asm>(
-     "cp.async.wait_group %0;",
+     "cp.async.wait_group",
...
```
The argument list is generated by `kir::Asm::parameters`, it not only supports primitive data types, which is converted to a plain item like `%1`, but also support arrays and pointers. Arrays will be expanded as things like `{%0, %1, %2, %3}`. Pointers will be converts to things like `[%1]`.

`codegen.cpp` is adjusted accordingly to accommodate this extension.

Thanks to this extension, we are capable of lowering `ldmatrix` as `kir::Asm`, and remove it from our runtime strings. `ldmatrix` itself is not long in the runtime strings, so this PR has more additions rather than deletions in terms of line of code, however, the extension on the support of arrays and pointers is a required step to lower MMA ops as `kir::Asm`, which is my final goal for this series of cleanups.

The code for ldmatrix taken from a generated matmul kernel:
```CUDA
    #pragma unroll
    for(nvfuser_index_t i88 = 0; i88 < 4; ++i88) {
      asm volatile(
        "ldmatrix.sync.aligned.x4.m8n8.shared.b16 {%0, %1, %2, %3}, [%4];\n"
        :"=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i88)]))[0]), 
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i88)]))[1]), 
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i88)]))[2]), 
         "=r"((*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T9[(8LL * i88)]))[3])
        :"r"((uint32_t)((i102 + (1024 * i88))))
      );
    }
    #pragma unroll
    for(nvfuser_index_t i89 = 0; i89 < 8; ++i89) {
      nvfuser_index_t i125;
      i125 = i43 + i89;
      asm volatile(
        "ldmatrix.sync.aligned.x2.trans.m8n8.shared.b16 {%0, %1}, [%2];\n"
        :"=r"((*reinterpret_cast<Array<uint32_t, 2, 1>*>(&T2[(4LL * i89)]))[0]), 
         "=r"((*reinterpret_cast<Array<uint32_t, 2, 1>*>(&T2[(4LL * i89)]))[1])
        :"r"((uint32_t)(((i103 + (128 * (i125 / 8))) + (16 * (i15 ^ (i125 % 8))))))
      );
    }
```

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/kernel_ir.cpp [hunk 1]
```diff
@@ -242,7 +242,27 @@ Asm::Asm(
...
+}
+
+const char* getPTXConstraints(Val* value) {
+  DataType dt = getTypeOrIndexType(value);
+  if (auto ti = dynamic_cast<kir::TensorIndex*>(value)) {
+    if (isPointerType(ti->index()->dtype())) {
+      dt = ti->index()->dtype();
+    }
+  }
+  if (std::holds_alternative<ArrayType>(dt.type)) {
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/kernel_ir.cpp]: The code here in intentional. I updated this PR and left some comments for it.

[the remark this answers is not in the document]
Comment (drzejan2) [reply csrc/kernel_ir.cpp]: Ok, thank you for adding documentation for this, as it wasn't clear given that `getTypeOrIndexType(...)` does exactly the same thing as the code that directly follows the call to this function.

**Pull request merged**